.design/requirements-575.md

```diff
@@ -0,0 +1,269 @@
+
+## Introduction
+
+The OSGi Compendium R6 introduced the HTTP Whiteboard specification, upgraded in R8.1 to the Whiteboard specification for Jakarta Servlet.
+It provides a light and convenient way of using servlets in a an OSGi environment through the use of the Whiteboard Pattern.
+
+WebSocket is communication protocol designed to work over the HTTP protocol.
+It enables bidirectional exchange of messages between the web client and the server.
+Like an HTTP servlet, a WebSocket handler is bound to a URL.
+
+The conversion from an HTTP connection to a WebSocket stream is done by sending a `GET` HTTP request with a set of headers requesting a connection upgrade.
+The server will respond with a *101 Switching protocols* response.
+
+The WebSocket protocol is defined by [RFC 6455](https://datatracker.ietf.org/doc/html/rfc6455) and a Java specification is defined in Jakarta EE.
+This document considers the specification will rely on [Jakarta WebSocket 2.1](https://jakarta.ee/specifications/websocket/2.1/), from Jakarta EE 10.
+
+HTTP server implementations can provide support for WebSocket, but their configuration is complex in the context of the OSGi Whiteboard specification for Jakarta Servlet.
+
+A Jakarta Websocket Whiteboard would ease the deployment of WebSocket handlers in an OSGi HTTP environment.
+This document describes the requirements for such a specification.
+
+## Terminology
+
+* *WebSocket* - The communication channel based on the WebSocket protocol.
+* *WebSocket frame* - Fragment of a WebSocket message as transferred on the WebSocket.
+* *WebSocket message* - Data or control message transferred over the WebSocket. A message can be transmitted via one or multiple frames. A data message can be text or binary. A control message can be a ping, pong or close message.
+* *WebSocket Container* - The server-side, implementation-specific, container of server WebSocket Endpoints.
+* *WebSocket Session* - Description of a single client-server WebSocket connection.
+* *WebSocket Endpoint* - Handler of a WebSocket session lifecycle: open, close, error. The endpoint registers frame listeners when the session is opened. Can be used either on client or server side.
+* *WebSocket (Message) Handler* - Handler notified of WebSocket message reception. Can be used either on client or server side.
+
+## Problem Description
+
+Registering a WebSocket handler can currently be done by using a custom configuration of an HTTP servlet.
+The problem is that such configuration depends on the underlying HTTP server implementation.
+
+For example, the Eclipse Jetty server requires registering components before initializing an HTTP servlet to allow it to handle the WebSocket protocol.
+This requires to write code specific to Jetty in an OSGi HTTP whiteboard environment.
+There is also an issue between the Jetty WebSocket server components and the servlet context given to it by the Apache Felix HTTP Jetty implementation.
```

Comment:
Jakarta Websockets include specific methods to interface with Sevlet API so these can be implemented in a container agnostic way.